File: Harness-UI/windows/HarnessStore.cs
using System.Text.Json;

namespace HarnessUI;

internal sealed class HarnessStore
{
    private readonly object gate = new();
    private readonly string dataRoot;
    private readonly JsonSerializerOptions json;
    private Catalog catalog = Catalog.Empty;
    private HarnessState state = new();
    private IReadOnlyDictionary<string, string> assets = new Dictionary<string, string>();

    internal HarnessStore(string dataRoot, JsonSerializerOptions json)
    {
        this.dataRoot = dataRoot;
        this.json = json;
        Directory.CreateDirectory(dataRoot);
        state = Read<HarnessState>(StateFile) ?? new HarnessState();
        catalog = Read<Catalog>(CatalogFile) ?? Catalog.Empty;
    }

    internal string CatalogFile => Path.Combine(dataRoot, "catalog.json");
    internal string StateFile => Path.Combine(dataRoot, "state.json");
    internal string ConfigFile => Path.Combine(dataRoot, "config.json");

    internal Catalog Catalog()
    {
        lock (gate) return catalog;
    }

    internal HarnessState State()
    {
        lock (gate) return Copy(state);
    }

    internal string? Asset(string requestPath)
    {
        lock (gate) return assets.TryGetValue(requestPath, out var file) ? file : null;
    }

    internal void Install(CatalogBuild build)
    {
        lock (gate)
        {
            catalog = build.Catalog;
            assets = build.Assets;
            state = Normalize(state, catalog);
            Write(CatalogFile, catalog);
            Write(StateFile, state);
        }
    }

    internal HarnessState Patch(StatePatch patch, long? now = null)
    {
        lock (gate)
        {
            if (patch.HasMode) state.Mode = patch.Mode ?? "gallery";
            if (patch.HasSelected) state.Selected = patch.Selected;
            if (patch.HasIntervalMs) state.IntervalMs = patch.IntervalMs;
            if (patch.HasHidden) state.Hidden = patch.Hidden;
            state.Updated = now ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            state = Normalize(state, catalog);
            Write(StateFile, state);
            return Copy(state);
        }
    }

    internal HarnessState Rotate(bool force, long? now = null)
    {
        lock (gate)
        {
            var timestamp = now ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            state = Normalize(state, catalog);
            if (state.Mode != "rotate") return Copy(state);
            if (!force && timestamp - state.LastRotate < state.IntervalMs) return Copy(state);

            if (state.Cycle.Length == 0 || state.Cursor >= state.Cycle.Length)
            {
                var hidden = state.Hidden.ToHashSet(StringComparer.Ordinal);
                state.Cycle = catalog.Entries.Select(entry => entry.Id).Where(id => !hidden.Contains(id)).ToArray();
                Random.Shared.Shuffle(state.Cycle);
                state.Cursor = 0;
            }
            if (state.Cursor >= state.Cycle.Length) return Copy(state);
            state.Selected = state.Cycle[state.Cursor++];
            state.LastRotate = timestamp;
            state.Updated = timestamp;
            Write(StateFile, state);
            return Copy(state);
        }
    }

    internal byte[] CatalogJson()
    {
        lock (gate) return JsonSerializer.SerializeToUtf8Bytes(catalog, json);
    }

    internal byte[] StateJson()
    {
        lock (gate) return JsonSerializer.SerializeToUtf8Bytes(state, json);
    }

    private HarnessState Normalize(HarnessState input, Catalog currentCatalog)
    {
        var ids = currentCatalog.Entries.Select(entry => entry.Id).ToHashSet(StringComparer.Ordinal);
        var hidden = input.Hidden.Where(ids.Contains).Distinct(StringComparer.Ordinal).Order(StringComparer.Ordinal).ToArray();
        var hiddenSet = hidden.ToHashSet(StringComparer.Ordinal);
        var cycle = input.Cycle.Where(id => ids.Contains(id) && !hiddenSet.Contains(id)).Distinct(StringComparer.Ordinal).ToArray();
        var selected = input.Selected;
        if (selected is null || !ids.Contains(selected) || hiddenSet.Contains(selected))
            selected = currentCatalog.Entries.FirstOrDefault(entry => !hiddenSet.Contains(entry.Id))?.Id;
        return input with
        {
            Mode = input.Mode == "rotate" ? "rotate" : "gallery",
            Selected = selected,
            IntervalMs = input.IntervalMs >= 60_000 ? input.IntervalMs : 14_400_000,
            Hidden = hidden,
            Cycle = cycle,
            Cursor = Math.Clamp(input.Cursor, 0, cycle.Length),
        };
    }

    private T? Read<T>(string file)
    {
        try { return JsonSerializer.Deserialize<T>(File.ReadAllBytes(file), json); }
        catch (IOException) { return default; }
        catch (JsonException) { return default; }
    }

    private void Write<T>(string file, T value)
    {
        var temporary = file + ".tmp";
        File.WriteAllBytes(temporary, JsonSerializer.SerializeToUtf8Bytes(value, json));
        File.Move(temporary, file, true);
    }

    private static HarnessState Copy(HarnessState value) => value with
    {
        Hidden = [.. value.Hidden],
        Cycle = [.. value.Cycle],
    };
}

File: Harness-UI/macos/Sources/HarnessUICore/HarnessStore.swift
import Foundation

public final class HarnessStore: @unchecked Sendable {
    private let queue = DispatchQueue(label: "com.linzecolin.harnessui.store")
    private var catalogValue: Catalog = .empty
    private var stateValue = HarnessState()
    private var assetsValue: [String: URL] = [:]
    private let dataRoot: URL
    private let encoder: JSONEncoder
    private let decoder = JSONDecoder()

    public init(dataRoot: URL) {
        self.dataRoot = dataRoot
        self.encoder = JSONEncoder()
        self.encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        try? FileManager.default.createDirectory(at: dataRoot, withIntermediateDirectories: true)
        if let data = try? Data(contentsOf: stateFile), let state = try? decoder.decode(HarnessState.self, from: data) {
            stateValue = state
        }
        if let data = try? Data(contentsOf: catalogFile), let catalog = try? decoder.decode(Catalog.self, from: data) {
            catalogValue = catalog
        }
    }

    public var catalogFile: URL { dataRoot.appendingPathComponent("catalog.json") }
    public var stateFile: URL { dataRoot.appendingPathComponent("state.json") }
    public var configFile: URL { dataRoot.appendingPathComponent("config.json") }

    public func catalog() -> Catalog { queue.sync { catalogValue } }
    public func state() -> HarnessState { queue.sync { stateValue } }
    public func asset(for requestPath: String) -> URL? { queue.sync { assetsValue[requestPath] } }

    public func install(build: CatalogBuild) throws {
        try queue.sync {
            if catalogValue.count > 0 && build.catalog.count == 0 {
                throw NSError(
                    domain: "HarnessUI",
                    code: 2,
                    userInfo: [NSLocalizedDescriptionKey: "素材源未返回任何有效皮肤；已保留上一次的 \(catalogValue.count) 个素材。请检查 SMB 连接或所选目录。"]
                )
            }
            let previousGames = Set(catalogValue.entries.map(\.game))
            let nextGames = Set(build.catalog.entries.map(\.game))
            let missingGames = previousGames.subtracting(nextGames)
            if !missingGames.isEmpty {
                throw NSError(
                    domain: "HarnessUI",
                    code: 3,
                    userInfo: [NSLocalizedDescriptionKey: "素材源缺少既有游戏分区：\(missingGames.sorted().joined(separator: ", "))；已保留上一版目录。"]
                )
            }
            catalogValue = build.catalog
            assetsValue = build.assets
            stateValue = normalized(stateValue, catalog: catalogValue)
            stateValue.catalogGenerated = build.catalog.generated
            stateValue.updated = Int64(Date().timeIntervalSince1970 * 1000)
            try persistLocked()
        }
    }

    public func patch(_ values: [String: Any], now: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) throws -> HarnessState {
        try queue.sync {
            if let mode = values["mode"] as? String { stateValue.mode = mode }
            if let selected = values["selected"] as? String { stateValue.selected = selected }
            if values["selected"] is NSNull { stateValue.selected = nil }
            if let interval = values["intervalMs"] as? NSNumber { stateValue.intervalMs = interval.intValue }
            if let hidden = values["hidden"] as? [String] { stateValue.hidden = hidden }
            stateValue.updated = now
            stateValue = normalized(stateValue, catalog: catalogValue)
            try persistStateLocked()
            return stateValue
        }
    }

    public func rotate(force: Bool, now: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) throws -> HarnessState {
        try queue.sync {
            stateValue = normalized(stateValue, catalog: catalogValue)
            guard stateValue.mode == "rotate" else { return stateValue }
            guard force || now - stateValue.lastRotate >= Int64(stateValue.intervalMs) else { return stateValue }
            if stateValue.cycle.isEmpty || stateValue.cursor >= stateValue.cycle.count {
                let hidden = Set(stateValue.hidden)
                stateValue.cycle = catalogValue.entries.map(\.id).filter { !hidden.contains($0) }.shuffled()
                stateValue.cursor = 0
            }
            guard stateValue.cursor < stateValue.cycle.count else { return stateValue }
            stateValue.selected = stateValue.cycle[stateValue.cursor]
            stateValue.cursor += 1
            stateValue.lastRotate = now
            stateValue.updated = now
            try persistStateLocked()
            return stateValue
        }
    }

    public func jsonCatalog() throws -> Data { try queue.sync { try encoder.encode(catalogValue) } }
    public func jsonState() throws -> Data { try queue.sync { try encoder.encode(stateValue) } }

    private func normalized(_ input: HarnessState, catalog: Catalog) -> HarnessState {
        var state = input
        state.catalogGenerated = catalog.generated.isEmpty ? state.catalogGenerated : catalog.generated
        let ids = Set(catalog.entries.map(\.id))
        state.mode = state.mode == "rotate" ? "rotate" : "gallery"
        if state.intervalMs < 60_000 { state.intervalMs = 14_400_000 }
        state.hidden = Array(Set(state.hidden.filter(ids.contains))).sorted()
        state.cycle = state.cycle.filter { ids.contains($0) && !state.hidden.contains($0) }
        state.cursor = max(0, min(state.cursor, state.cycle.count))
        if let selected = state.selected, !ids.contains(selected) || state.hidden.contains(selected) { state.selected = nil }
        if state.selected == nil { state.selected = catalog.entries.first?.id }
        return state
    }

    private func persistLocked() throws {
        try encoder.encode(catalogValue).write(to: catalogFile, options: .atomic)
        try persistStateLocked()
    }

    private func persistStateLocked() throws {
        try encoder.encode(stateValue).write(to: stateFile, options: .atomic)
    }
}
